analog_reader.go

```diff
@@ -26,82 +27,77 @@ type AnalogReader interface {
 	Range(steps uint16) uint16
 }
 
-// A struct for handling the reading of analogue control voltage.
-// The analogue input allows you to 'read' CV from anywhere between 0 and 12V.
-type AnalogInput struct {
+type analogInput struct {
 	machine.ADC
 	samples uint16
 }
 
-// NewAI creates a new AnalogInput.
-func NewAI(pin machine.Pin) *AnalogInput {
+func newAnalogInput(pin machine.Pin) *analogInput {
 	adc := machine.ADC{Pin: pin}
 	adc.Configure(machine.ADCConfig{})
-	return &AnalogInput{ADC: adc, samples: DefaultSamples}
+	return &analogInput{ADC: adc, samples: defaultSamples}
 }
 
 // Samples sets the number of reads for an more accurate average read.
-func (a *AnalogInput) Samples(samples uint16) {
+func (a *analogInput) Samples(samples uint16) {
 	a.samples = samples
 }
 
 // Percent return the percentage of the input's current relative range as a float between 0.0 and 1.0.
-func (a *AnalogInput) Percent() float32 {
-	return float32(a.read()) / CalibratedMaxAI
+func (a *analogInput) Percent() float32 {
+	return float32(a.read()) / calibratedMaxAI
 }
 
 // ReadVoltage return the current read voltage between 0.0 and 10.0 volts.
-func (a *AnalogInput) ReadVoltage() float32 {
+func (a *analogInput) ReadVoltage() float32 {
 	return a.Percent() * MaxVoltage
 }
 
 // Range return a value between 0 and the given steps (not inclusive) based on the range of the analog input.
-func (a *AnalogInput) Range(steps uint16) uint16 {
+func (a *analogInput) Range(steps uint16) uint16 {
 	return uint16(a.Percent() * float32(steps))
 }
 
-func (a *AnalogInput) read() uint16 {
+func (a *analogInput) read() uint16 {
 	var sum int
 	for i := 0; i < int(a.samples); i++ {
-		sum += Clamp(int(a.Get())-CalibratedMinAI, 0, CalibratedMaxAI)
+		sum += Clamp(int(a.Get())-calibratedMinAI, 0, calibratedMaxAI)
 	}
 	return uint16(sum / int(a.samples))
 }
 
-// A struct for handling the reading of knob voltage and position.
-type Knob struct {
+type knob struct {
 	machine.ADC
```

Comment:
Careful with exposing a `machine.ADC` here. I'd unexport this field or switch to an interface. This will prevent from users depending on tinygo directly and make mock-tests possible.

Reply:
I'll ditto that remark - in my fork of the project I've been fighting against the lack of testability quite often because things are so intrinsically linked to `machine` and other TinyGo packages.

My main personal failing until now is that I've just been building runtime testing apparatus frameworks instead of just abstracting away the TinyGo linkages.... though those don't hurt to have around for final-pass QA, I guess.